File: server/open.ts
import { spawn } from "node:child_process";

export type OpenTarget = "vscode" | "cursor" | "terminal" | "iterm" | "finder" | "url";

/**
 * Launch an external app/URL. `arg` is a directory path (editors/terminal/finder)
 * or a URL ("url"). Resolves when the launcher exits 0, rejects with its error
 * (e.g. the editor CLI isn't installed) so the UI can surface it.
 */
export function openTarget(target: OpenTarget, arg: string): Promise<void> {
  const plat = process.platform;
  let cmd: string;
  let args: string[];

  switch (target) {
    case "url":
      cmd = plat === "darwin" ? "open" : plat === "win32" ? "explorer" : "xdg-open";
      args = [arg];
      break;
    case "finder":
      cmd = plat === "darwin" ? "open" : plat === "win32" ? "explorer" : "xdg-open";
      args = [arg];
      break;
    case "terminal":
      if (plat === "darwin") {
        cmd = "open";
        args = ["-a", "Terminal", arg];
      } else if (plat === "win32") {
        cmd = "cmd";
        args = ["/c", "start", "cmd", "/K", `cd /d ${arg}`];
      } else {
        cmd = "x-terminal-emulator";
        args = [];
      }
      break;
    case "iterm":
      cmd = "open";
      args = ["-a", "iTerm", arg];
      break;
    case "vscode":
      cmd = "code";
      args = [arg];
      break;
    case "cursor":
      cmd = "cursor";
      args = [arg];
      break;
    default:
      return Promise.reject(new Error(`Unknown open target: ${target}`));
  }

  return new Promise((resolve, reject) => {
    const child = spawn(cmd, args, { stdio: "ignore" });
    child.on("error", (e: NodeJS.ErrnoException) =>
      reject(new Error(e.code === "ENOENT" ? `${cmd} is not installed or not on PATH` : e.message)),
    );
    child.on("exit", (code) => (code === 0 || code === null ? resolve() : reject(new Error(`${cmd} exited with code ${code}`))));
  });
}

File: src-tauri/src/open.rs
//! External app/URL launcher — mirrors server/open.ts. std-only (no tauri) so it
//! compiles into the headless `kablan-server` binary as well as the desktop app.
use std::io::ErrorKind;
use std::process::Command;

/// Launch an external app/URL. `arg` is a directory path (editors/terminal/finder)
/// or a URL ("url"). Returns Ok(()) when the launcher exits 0 (or with no code),
/// or Err with its error (e.g. the editor CLI isn't installed) so the UI can surface it.
pub fn open_target(target: &str, arg: &str) -> Result<(), String> {
    let plat = std::env::consts::OS;
    let (cmd, args): (&str, Vec<String>) = match target {
        "url" | "finder" => {
            let cmd = match plat {
                "macos" => "open",
                "windows" => "explorer",
                _ => "xdg-open",
            };
            (cmd, vec![arg.to_string()])
        }
        "terminal" => match plat {
            "macos" => ("open", vec!["-a".into(), "Terminal".into(), arg.to_string()]),
            "windows" => (
                "cmd",
                vec![
                    "/c".into(),
                    "start".into(),
                    "cmd".into(),
                    "/K".into(),
                    format!("cd /d {arg}"),
                ],
            ),
            _ => ("x-terminal-emulator", vec![]),
        },
        "iterm" => ("open", vec!["-a".into(), "iTerm".into(), arg.to_string()]),
        "vscode" => ("code", vec![arg.to_string()]),
        "cursor" => ("cursor", vec![arg.to_string()]),
        _ => return Err(format!("Unknown open target: {target}")),
    };

    let status = Command::new(cmd).args(&args).status();
    match status {
        Ok(s) => {
            if s.success() {
                Ok(())
            } else {
                match s.code() {
                    Some(code) => Err(format!("{cmd} exited with code {code}")),
                    None => Ok(()),
                }
            }
        }
        Err(e) => {
            if e.kind() == ErrorKind::NotFound {
                Err(format!("{cmd} is not installed or not on PATH"))
            } else {
                Err(e.to_string())
            }
        }
    }
}
